Extend PgBackendStatus with waiting event start time.

InnerLife0 opened the pull request:
As we want to somehow monitor main waiting events (PGBE_WAITING_LOCK, PGBE_WAITING_REPLICATION) we need to add a time events initially started. 
We also need to clear it once awaiting is over (PGBE_WAITING_NONE). To handle this, we add new st_waiting_start_timestamp field and working with it when track_activities GUC is enabled.

maksm90 commented on src/backend/postmaster/pgstat.c:
```diff
@@ -2913,6 +2915,10 @@ gpstat_report_waiting(char reason)
+	if (reason == PGBE_WAITING_NONE)
+		beentry->st_waiting_start_timestamp = 0;
+	else
+		beentry->st_waiting_start_timestamp = GetCurrentTimestamp();
```
Does set up timestamp just regard to waiting on locks? Here we track the starting timestamp for all wait events. So commit message and PR description have to be revised.

InnerLife0 replied:
Fixed. Thanks.

Stolb27 commented on src/backend/postmaster/pgstat.c:
```diff
@@ -2913,6 +2915,10 @@ gpstat_report_waiting(char reason)
 	 * protocol.  The update must appear atomic in any case.
```
It seems that this comment isn't actual after this patch. Why didn't you wrap changes with the st_changecount protocol?

InnerLife0 replied:
Fixed. Thanks.

maksm90 commented on src/backend/postmaster/pgstat.c:
```diff
@@ -2908,11 +2910,19 @@ gpstat_report_waiting(char reason)
-	 * may modify, there seems no need to bother with the st_changecount
-	 * protocol.  The update must appear atomic in any case.
+	 * Since we updating multiple fields, go through the st_changecount
+	 * protocol.
```
the canonical phrase `protocol of bumping st_changecount before and after` is more relevant here

InnerLife0 replied:
![Webp net-resizeimage](https://user-images.githubusercontent.com/42867699/162230921-30e6679b-7108-4335-a60c-151dd44d6580.jpg)

Commit:
Extend PgBackendStatus with waiting event start time.
As we want to somehow monitor main waiting events (PGBE_WAITING_LOCK, PGBE_WAITING_REPLICATION) we need to add a time events initially started.
We also need to clear it once awaiting is over (PGBE_WAITING_NONE). To handle this, we add new st_waiting_start_timestamp field and working with it when track_activities GUC is enabled.
